modules/garbage.py

```diff
@@ -2,124 +2,108 @@
 
     def run(self):
-        try:
-            while True:
-                if self.heap.level >= self.threshold:
-                    self.env.process(self.collect())
-                yield self.env.timeout(self.sleep)
+        while True:
+            if self.heap.level >= self.threshold and not self.collect_exe:
+                yield self.env.process(self.collect())
+                self.times_performed += 1
 
-        except simpy.Interrupt:
             yield self.env.timeout(self.sleep)  # wait for...
 
-
     def collect(self):
-        print("At %.3f, CGC GC is running. We have %.3f of trash" % (self.env.now, self.heap.level))
-
+        print("At %.3f, GC is running. We have %.3f of trash" % (self.env.now, self.heap.level))
+        self.collect_exe = True
+        self.server.action.interrupt()
         while self.heap.level > 0:                                          # while threshold is empty...
             trash = self.heap.level                                         # keeps the amount of trash
             yield self.env.timeout(self.gc_execution_time_by_trash(trash))  # run the time of discarting
             yield self.heap.get(trash)                                      # discards the trash
 
-        print("At %.3f, CGC GC finish his job. Now we have %.3f of trash" % (self.env.now, self.heap.level))
+        self.collect_exe = False
+        print("At %.3f, GC finish his job. Now we have %.3f of trash" % (self.env.now, self.heap.level))
+        self.server.action = self.env.process(self.server.run())
 
     def gc_execution_time_by_trash(self, trash):
         """ implement the way to calculate the execution time of Garbage Collector """
         return trash
 
-class STWGC(GC):
-
-    def __init__(self, env, server, threshold=1, sleep=2):
-        super().__init__(env, server.heap, threshold, sleep)
-        self.server = server
-
-    def collect(self):
-        print("At %.3f, STWGC GC is running. We have %.3f of trash" % (self.env.now, self.heap.level))
-
-        self.server.action.interrupt()
-        if self.server.gc.collect_exe:
-            self.server.gc.collect_exe.interrupt()
-
-        while self.heap.level > 0:                                          # while threshold is empty...
-            trash = self.heap.level                                         # keeps the amount of trash
-            yield self.env.timeout(self.gc_execution_time_by_trash(trash))  # run the time of discarting
-            yield self.heap.get(trash)                                      # discards the trash
-
-        self.server.action = self.env.process(self.server.run())
-
-        print("At %.3f, STWGC GC finish his job. Now we have %.3f of trash" % (self.env.now, self.heap.level))
 
 class GCI(object):
 
-    def __init__(self, env,  threshold=0.7, check_heap=2, initial_gc_exec_time=0.200):
+    def __init__(self, env, server,  threshold=0.7, check_heap=2, initial_gc_exec_time=0, sleep=0.00002):
```

Comment:
why do we need the sleep variable?

Reply:
and what `initial_gc_exec_time` mean?

Reply:
The sleep variable is used to create a delay after each execution if needed. I believe the name of this variable is not good too...
The initial_gc_exec_time variable was created thinking in represent the inicial time cost for start GCI. I believe it is not being used on the current model and it will probably be removed.

Reply:
This is the initial estimated execution time of GC, this was think to cover the initial case when GCI doesn't have any history of GC execution time.

Reply:
Name changed to "initial_estimated_gc_exec_time". Could you review?

Reply:
Please remove altogether. Let's get rid of unnecessary complexity. There is no (negligible) cost in creating GCI. GCI is a plain **function call** :smile:

Reply:
Ok. we'll remove it.

Reply:
@danielfireman if we remove this completely how we will estimate the execution time of the first GC execution?